bundles/org.openhab.binding.openthermgateway/src/main/java/org/openhab/binding/openthermgateway/handler/OpenThermGatewayHandler.java

```diff
@@ -108,183 +101,184 @@ public void handleCommand(ChannelUID channelUID, Command command) {
-            logger.debug("Scheduling to reconnect in {} seconds.", conf.connectionRetryInterval);
-            reconnectTask = scheduler.schedule(this::connect, conf.connectionRetryInterval, TimeUnit.SECONDS);
+            if (handler != null) {
+                handler.receiveMessage(message);
+            }
         }
     }
 
     @Override
-    public void receiveMessage(Message message) {
-        if (DataItemGroup.dataItemGroups.containsKey(message.getID())) {
-            DataItem[] dataItems = DataItemGroup.dataItemGroups.get(message.getID());
-
-            for (DataItem dataItem : dataItems) {
-                String channelId = dataItem.getSubject();
-
-                if (!OpenThermGatewayBindingConstants.SUPPORTED_CHANNEL_IDS.contains(channelId)
-                        || (dataItem.getFilteredCode() != null && dataItem.getFilteredCode() != message.getCode())) {
-                    continue;
-                }
-
-                State state = null;
-
-                switch (dataItem.getDataType()) {
-                    case FLAGS:
-                        state = OnOffType.from(message.getBit(dataItem.getByteType(), dataItem.getBitPos()));
-                        break;
-                    case UINT8:
-                    case UINT16:
-                        state = new DecimalType(message.getUInt(dataItem.getByteType()));
-                        break;
-                    case INT8:
-                    case INT16:
-                        state = new DecimalType(message.getInt(dataItem.getByteType()));
-                        break;
-                    case FLOAT:
-                        float value = message.getFloat();
-                        @Nullable
-                        Unit<?> unit = dataItem.getUnit();
-                        state = (unit == null) ? new DecimalType(value) : new QuantityType<>(value, unit);
-                        break;
-                    case DOWTOD:
-                        break;
-                }
+    public void connectionStateChanged(ConnectionState state) {
+        scheduler.submit(() -> connectionStateChangedTask(state));
+    }
 
-                if (state != null) {
-                    logger.debug("Received update for channel '{}': {}", channelId, state);
-                    updateState(channelId, state);
-                }
+    private void connectionStateChangedTask(ConnectionState state) {
+        if (this.state != state) {
+            this.state = state;
+
+            switch (state) {
+                case INITIALIZING:
+                    updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.NONE, "Initializing");
+                    break;
+                case DISCONNECTED:
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, "Disconnected");
+                    autoReconnect();
+                    break;
+                case CONNECTING:
+                    updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.NONE, "Connecting");
```

Comment:
What could be the reason for this ConnectionState? Initially **UNKNOWN** is correct, but if this can happen later because device is offline and we now need to reconnect, **OFFLINE** would be more appropriate. Didn't check how it's used.

Reply:
This is a callback which is called several times during the various stages of establishing a TCP socket connection to the gateway device. 

I suppose one could argue that if the Bridge is not specifically ONLINE then it is _per definition_ OFFLINE. In that case the switch `case CONNECTED` could return ONLINE and the switch `default` case should return OFFLINE. So in other words the switch statement could be simplified to the following. PS also removes the need for special i18n texts :)

```
            switch (state) {
                case CONNECTED:
                    updateStatus(ThingStatus.ONLINE);
                    cancelAutoReconnect();
                    break;
                case DISCONNECTED:
                    autoReconnect();
                    // fall through
                default:
                    updateStatus(ThingStatus.OFFLINE);
            }
```

Reply:
^
Probably the following is even better than the above..
```
            switch (state) {
                case CONNECTED:
                    updateStatus(ThingStatus.ONLINE);
                    cancelAutoReconnect();
                    break;
                case DISCONNECTED:
                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
                    autoReconnect();
                    break;
                case CONNECTING:
                case INITIALIZING:
                    updateStatus(ThingStatus.INITIALIZING);
            }
```

Reply:
Actually `case DISCONNECTED` can occur both a) if there was a real communication error and b) if the Bridge was disabled by the framework; so strictly in the latter case, `ThingStatusDetail.COMMUNICATION_ERROR` is not correct a correct statement; but I think it doesn't really matter because the framework will set 'ThingStatusDetail.DISABLED' immediately afterwards. Or ??

Reply:
Not sure, sounds plausible, but probably you would need to test it.

Reply:
> you would need to test it
Ok, I have tested it now :) and I found out that setting `ThingStatus.INITIALIZING` is not allowed, so I have pushed a (hopefully) final commit that (hopefully) closes this topic.